### PR TITLE
 [ParseableInterface] Serialize module into buffer and load from it

### PR DESCRIPTION
**Pull request by harlanhaskins**
In addition to being wasteful, this is a correctness issue -- the
compiler should only ever have one view of this file, and it should not
read a potentially different file after validating dependencies.

rdar://48654608

### CODE REVIEW

**Comment by Azoy** [comment]
Really amazing comments in this patch!

**Comment by harlanhaskins** [comment]
@Azoy Thanks!

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: lib/Frontend/ParseableInterfaceModuleLoader.cpp [hunk 1]
```diff
@@ -173,30 +173,40 @@ class DiscoveredModule {
...
+  std::unique_ptr<llvm::MemoryBuffer> moduleBuffer;
+
   /// The path to the discovered serialized .swiftmodule on disk.
   const std::string path;
 
   /// Creates a \c Normal discovered module.
-  static DiscoveredModule normal(StringRef path) {
-    return { path, Kind::Normal };
+  static DiscoveredModule normal(StringRef path,
+    std::unique_ptr<llvm::MemoryBuffer> moduleBuffer) {
```
Comment (jrose-apple) [inline lib/Frontend/ParseableInterfaceModuleLoader.cpp]: Nitpick: double-indent wrapped lines (per clang-format).

File: lib/Serialization/Serialization.cpp [hunk 2]
```diff
@@ -4809,6 +4810,48 @@ void Serializer::writeToStream(raw_ostream &os, ModuleOrSourceFile DC,
   S.writeToStream(os);
 }
 
+void swift::serializeToBuffers(
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: Can you implement `serialize` in terms of this, just so we don't have two implementations?

File: lib/Serialization/Serialization.cpp [hunk 3]
```diff
@@ -48,6 +48,7 @@
 #include "llvm/Support/EndianStream.h"
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SmallVectorMemoryBuffer.h"
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: Nitpick: alphabet

File: lib/Frontend/ParseableInterfaceModuleLoader.cpp [hunk 4]
```diff
@@ -559,7 +570,14 @@ class swift::ParseableInterfaceBuilder {
       if (ShouldSerializeDeps)
         SerializationOpts.Dependencies = Deps;
       SILMod->setSerializeSILAction([&]() {
-        serialize(Mod, SerializationOpts, SILMod.get());
+        if (ModuleBuffer) {
+          // We don't want to serialize module docs in the cache -- they
+          // will be serialized beside the interface file.
+          serializeToBuffers(Mod, SerializationOpts, ModuleBuffer,
+                             /*ModuleDocBuffer*/nullptr, SILMod.get());
```
Comment (jrose-apple) [inline lib/Frontend/ParseableInterfaceModuleLoader.cpp]: Seems like we could always pass down `ModuleBuffer` and `nullptr` and get the same effect?

File: include/swift/Subsystems.h [hunk 5]
```diff
@@ -264,6 +264,17 @@ namespace swift {
   void serialize(ModuleOrSourceFile DC, const SerializationOptions &options,
                  const SILModule *M = nullptr);
 
+  /// Serializes a module or single source file to the given output file and
+  /// returns back the file's contents as a memory buffer.
+  /// Use this if you intend to immediately load the serialized module, as that
```
Comment (jrose-apple) [inline include/swift/Subsystems.h]: Docs nitpick: this would be a good place for a newline to separate the "short" and "full" docs.

**Review by jrose-apple — COMMENTED** [review]
I like the safety of `std::unique_ptr` a lot here. Ownership FTW!

File: lib/Frontend/ParseableInterfaceModuleLoader.cpp (hunk 4 above)
Comment (harlanhaskins) [reply lib/Frontend/ParseableInterfaceModuleLoader.cpp]: I thought about that, but I preferred explicitly encoding the difference in behavior. But if you think they should be combined, I'll combine them.

File: lib/Serialization/Serialization.cpp (hunk 2 above)
Comment (harlanhaskins) [reply lib/Serialization/Serialization.cpp]: Are we cool with double-buffering the serialization?

Comment (jrose-apple) [reply lib/Serialization/Serialization.cpp]: Oh, right. It won't matter once we get a chance to do the mmap thing, but it does now.

File: lib/Frontend/ParseableInterfaceModuleLoader.cpp (hunk 4 above)
Comment (jrose-apple) [reply lib/Frontend/ParseableInterfaceModuleLoader.cpp]: I feel like the less a function changes its behavior based on inputs, the better.

File: lib/Serialization/Serialization.cpp [hunk 6]
```diff
@@ -4839,7 +4839,7 @@ void swift::serializeToBuffers(
     SharedTimer timer("Serialization, swiftdoc, to buffer");
     llvm::SmallString<1024> buf;
     llvm::raw_svector_ostream stream(buf);
-    Serializer::writeToStream(stream, DC, M, options);
+    writeDocToStream(stream, DC, options.GroupInfoPath);
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: Heh, good catch.

**Review by jrose-apple — APPROVED** [review]

**Review by benlangmuir — APPROVED** [review]
LGTM, thanks Harlan!

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test Linux

File: lib/ClangImporter/ClangImporter.cpp [hunk 7]
```diff
@@ -571,28 +600,10 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
...
-    // Only specify the module map if that file actually exists.
-    // It may not--for example in the case that
-    // `swiftc -target x86_64-unknown-linux-gnu -emit-ir` is invoked using
-    // a Swift compiler not built for Linux targets.
-    if (llvm::sys::fs::exists(GlibcModuleMapPath)) {
-      invocationArgStrs.push_back(
-        (Twine("-fmodule-map-file=") + GlibcModuleMapPath).str());
+    if (auto path = getGlibcModuleMapPath(searchPathOpts.RuntimeResourcePath,
+                                          triple, GlibcModuleMapPath)) {
+      invocationArgStrs.push_back(*path);
```
Comment (jrose-apple) [inline lib/ClangImporter/ClangImporter.cpp]: You lost the `-fmodule-map-file=`.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test Linux

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test macOS

**Comment by harlanhaskins** [comment]
Seems like the Linux bot had issues. Let's try again.

@swift-ci please smoke test Linux

**Review by jrose-apple — APPROVED** [review]

**Comment by harlanhaskins** [comment]
@swift-ci please test and merge

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Pull request merged**